### ISSUE TITLE
Clarify HTTPConnectionPool.getConnection dosctring

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6153]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#6153|
|Type|defect|
|Created|2012-10-22 18:14:44Z|
|Branch|https://github.com/twisted/twisted/tree/getconnection-docstring-6153|

The current docstring is a little confusing.

```
[10:39:19] <cyli> Am I not parsing this correctly (docstring for HTTPConnectionPool)?  "If the connection doesn't disconnect at the end of its request, it will be returned to this pool automatically.  As such, only a single request should be sent using the returned connection."  Doesn't the second part of that implies that the first part is missing a "not"
[10:40:28] <cyli> Er, sorry, HTTPConnectionPool.getConnection
[10:40:44] <dreid> cyli: The caller isn't responsible for adding the connection back to the pool.
[10:42:59] <cyli> dreid: so it's basically saying "don't terminate the connection, because then it won't be added back into the connection pool"?
[10:43:14] <dreid> So you need to call getConnection for every request.  You may get the same connection back from multiple calls to getConnection, but you can't make more than one request for each call to getConnection because it will be added back to the pool when the request finishes.
[10:44:32] <dreid> cyli: No, it's saying "Don't cache connections yourself"
[10:44:52] <cyli> dreid: It's not saying that very clearly
[10:45:32] <dreid> cyli: I'm inclined to agree.
[10:45:33] <cyli> dreid:  Actually, it's not saying that at all.  It may be implying it, obliquely.
[10:46:42] <cyli> dreid: Ok, thanks.  I didn't see how the first sentence and second sentence related, but your explanation makes much more sense.
[10:47:01] <dreid> cyli: They're related by "As such" :)
[10:47:49] <cyli> dreid: yeah that's what confused me, I didn't understand why the second followed the first.
[11:01:45] <glyph> cyli: Sounds bug-worthy
[11:02:11] <glyph> dreid's sentence should go in the docs I think
```

Attachments:

* [ticket-6153-docs.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2fa/2fa7b2908dcf473bba5da6d5755f13b6dbc531e7/76e353dbde721f68af42de97cda7f42e4b5e3c67.patch) (868 bytes) - added by wirehead on 2013-03-18 22:28:49Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6153 6153
type__defect defect
reporter__cyli cyli
priority__normal normal
milestone__ 
branch__branches_getconnection_docstring_6153 branches/getconnection-docstring-6153
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1350929684000000 1350929684000000
changetime__1365196002000000 1365196002000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="wirehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|wirehead set owner to wirehead<br>wirehead set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="wirehead's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wirehead removed owner<br>wirehead set status to `new`|
|-|-|

tried to figure out if there was a better way to say it, failed, copied the sentence in.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince commented|
|-|-|

How about "When a request is finished, the connection is returned to the pool if it hasn't been disconnected. As such, ...."

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [37935]) Branching to 'getconnection-docstring-6153'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [37936]) Apply ticket-6153-docs.patch 

refs [#6153](https://github.com/twisted/twisted/issues/6153)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Thanks for the suggestions.  I still didn't find the docstring all that clear, so I gave it a try myself.  My strategy here is generally to simplify sentence structure, avoid negations, and consider the big picture.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_8"></a>|@jonathanj set owner to @exarkun|
|-|-|

The simplified documentation does a good job of describing what the underlying process is and what the consequences are of making multiple requests over the same connection, which means the caller can decide if that is acceptable to them or not. A definite improvement in clarity.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [37948]) Merge getconnection-docstring-6153

Author: wirehead, exarkun
Reviewer: jonathanj
Fixes: [#6153](https://github.com/twisted/twisted/issues/6153)

Tweak the `twisted.web.client.HTTPConnectionPool.getConnection` docstring to try
to make it more clear what you should do with the connection you get from it.
